### ISSUE TITLE
Quarkus cache CompositeCacheKey only consists of method arguments - should include method name as well

### ISSUE THREAD

**Issue by joerg-db** [quarkusio/quarkus#23515]
### Description

Currently, the CompositeCacheKey of the Quarkus cache (caffeine) extension is calculating a deep hash on the list of method arguments. This is not sufficient if there are multiple different methods with the same signature.

Sample :  method getXYZ(String x, String z) and method getABC(String x, String z) would result in conflicting/equal CompositeCacheKeys leading to unexpected behaviour (e.g. wrong results are returned).

### Implementation ideas

Incl. the current method name as part of the CompositeCacheKey instead of relying on different argument names,

**Comment by gwenneg** [comment]
Hi @joerg-db.

If the method name was included in the cache key, then in the following code `invalidate` would no longer be able to invalidate the data cached with `load`.

```
@CacheResult(cacheName = "my-cache")
public String load(String keyElement1, Object keyElement2) { 
    // Call expensive service here.
}

@CacheInvalidate(cacheName = "my-cache")
public void invalidate(String keyElement1, Object keyElement2) { 
}
```

The "collision" between the methods cached data because of identical keys (composite or not) is actually intentional. If you need to make sure the values from 2 methods are not mixed in the cache, I would advise to use a different `cacheName` field value in the `@CacheResult` annotations.

**Comment by joerg-db** [comment]
Thx - that could be an option, but I would end up with lots of caches then. I would also suggest to clearly describe that behaviour in the docs, as it came totally unexpected to me that results of different methods are mixing up in the cache. Only found out during my unit tests.
Btw, during my test I found that [quarkus.cache.caffeine."cache-name".metrics-enabled](https://quarkus.io/guides/all-config#quarkus-cache_quarkus.cache.caffeine.-cache-name-.metrics-enabled) is not recognized as a valid config option.

**Comment by gwenneg** [comment]
> one more question - is the name of the argument making any difference ? I could rename the arguments much easier than setting up new caches

No the cache key is built from the values only, the names are ignored.

**Comment by gwenneg** [comment]
> Thx - that could be an option, but I would end up with lots of caches then. I would also suggest to clearly describe that behaviour in the docs, as it came totally unexpected to me that results of different methods are mixing up in the cache. Only found out during my unit tests.

Good point, I'll update the doc.

**Comment by gwenneg** [comment]
> Btw, during my test I found that [quarkus.cache.caffeine."cache-name".metrics-enabled](https://quarkus.io/guides/all-config#quarkus-cache_quarkus.cache.caffeine.-cache-name-.metrics-enabled) is not recognized as a valid config option.

That's weird, it works in our [integration test](https://github.com/quarkusio/quarkus/blob/b8ad9658f99026d80a67805b6d4b44d20dbd34c3/integration-tests/cache/src/main/resources/application.properties#L9) AFAIK. I'll check that.

**Comment by gsmet** [comment]
> Btw, during my test I found that [quarkus.cache.caffeine."cache-name".metrics-enabled](https://quarkus.io/guides/all-config#quarkus-cache_quarkus.cache.caffeine.-cache-name-.metrics-enabled) is not recognized as a valid config option.

Can you open an issue for this with a reproducer?

> I could rename the arguments much easier than setting up new caches

I wonder if we should have some sort of templating feature for the caches. E.g. you could have a default template for additional caches config. And you could also have named templates that you could point caches to.

Does it make any sense? /cc @gwenneg

**Comment by gsmet** [comment]
Yeah so that's normal. It's a 2.7 feature :).

**Comment by joerg-db** [comment]
or am I missing a required extension to get these metrics enabled ?

**Comment by joerg-db** [comment]
> Yeah so that's normal. It's a 2.7 feature :).

ok, understood - thx.

Coming back to the "CompositeCacheKey" problem for methods with same signature.

How about introducing a new optional field in the @CacheResult and corresponding annotations (like @CacheInvalidate) to give a suffix to the computed composite key ?

Something similar to 
@CacheResult(cache-name='mycache', cache-key-suffix='mysuffix')

That would make the keys unique per method just by using one cache configuration.

**Comment by joerg-db** [comment]
just to recap what options exist to overcome any composite key conflicts for methods having the same signature
1) Use different caches and select the proper cache by name : I would end up with many caches actually having the same eviction config just to work around the key clashes. Not very elegant imo
2) I could introduce a further argument to my methods that would act as a key name : also not very elegant, as this argument is actually not needed for my method
3) Introduce an optional annotation field "cache-key-name" to distinguish keys from methods with same signature :  that seems to be the most elegant way imo

Any thoughts ?

**Comment by gwenneg** [comment]
`3` is an interesting option but if we go that way, I would prefer a more flexible key builder/generator that would be implemented by the app developer from an interface provided by `quarkus-cache`. The args could be the method name, arguments, return type, etc.

`1` and `2` are the two options that can be used immediately indeed.

**Comment by gwenneg** [comment]
```
public class MyKeyBuilder implements KeyBuilder {
    @override
    public Object build(String methodName, Object... methodParams) { // args to be refined
        // compute and return the key
    }
}

@ApplicationScoped
public class Cached {

    @CacheResult(cacheName = "foo", keyBuilder = MyKeybuilder.class)
    public String bar() {
        // return something
    }

    @CacheInvalidate(cacheName = "foo", keyBuilder = MyKeybuilder.class)
    public void clearCache() {
    }
}
```

**Comment by gwenneg** [comment]
> I wonder if we should have some sort of templating feature for the configuration of the caches. E.g. you could have a default template for additional caches config. And you could also have named templates that you could point caches to.

Are you thinking of a template for cache keys or something that would only cover the configuration from `application.properties` or both at the same time? 😃

**Comment by gwenneg** [comment]
Here's a quick and dirty [example](https://github.com/gwenneg/quarkus/blob/bf22a2c6b3e2deeada87a3e99b5c5cec05aab07e/extensions/cache/deployment/src/test/java/io/quarkus/cache/test/runtime/CacheKeyGeneratorTest.java) of what I described above and the corresponding [PR](https://github.com/gwenneg/quarkus/pull/209) on my fork. It may not build, I improvised that super quickly 😄

**Comment by joerg-db** [comment]
I looked at your sample code and understood how a solution could be built.
Introducing a new KeyBuilder interface that let an app developer have control over how a composite cache key is actually built is definitely much more flexible than what I've suggested.

Looking forward for such a change to be released as I think it would be a real useful enhancement.

Just as a side note, what I'm currently trying to do is to implement a local JAX-RS stub for a remote microprofile API that basically has 20+ methods with (String a, String b) arguments - a nightmare with the current cache implementation ;-)

Thx for your valuable help and suggestions.
